src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+/*-------------------------------------------------------------------------
+ *
+ * dependency.c
+ *	  Functions to reason about distributed objects and their dependencies
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#include "access/skey.h"
+#include "catalog/dependency.h"
+#include "catalog/indexing.h"
+#include "catalog/pg_depend.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+
+#include "distributed/dist_catalog/dependency.h"
+#include "distributed/dist_catalog/distobject.h"
+
+
+static bool SupportedDependencyByCitus(const ObjectAddress *address);
+static bool IsObjectAddressInList(const ObjectAddress *findAddress, List *addressList);
+static bool IsObjectAddressOwnedByExtension(const ObjectAddress *target);
+
+static void recurse_pg_depend(const ObjectAddress *target,
+							  bool (*follow)(void *context, const Form_pg_depend row),
+							  void (*apply)(void *context, const Form_pg_depend row),
+							  void *context);
+static bool follow_order_object_address(void *context, const Form_pg_depend pg_depend);
+static bool follow_get_dependencies_for_object(void *context,
+											   const Form_pg_depend pg_depend);
+static void apply_add_to_target_list(void *context, const Form_pg_depend pg_depend);
+
+
+/*
+ * GetDependenciesForObject returns a list of ObjectAddesses to be created in order
+ * before the target object could safely be created on a worker. Some of the object might
+ * already be created on a worker. It should be created in an idempotent way.
+ */
+List *
+GetDependenciesForObject(const ObjectAddress *target)
+{
+	List *dependencyList = NIL;
```

Comment:
Super minor and could be personal, but I like to read with empty lines between the variable declarations, code and the return.
```C
List *dependencyList = NIL;

recurse...
...
...(..);

return dependencyList;
```

Reply:
Another convention note: We try to define the functions that are called within a function right after the function itself. So, in that sense, `recurse_pg_depend ` could be the next function.

In general, you might need to re-order the functions a bit.

We're not applying this convention very strictly, but it makes it slightly easier to read the code. So, you can make the call to re-order the functions or not.